pinot-core/src/main/java/com/linkedin/pinot/core/common/datatable/DataTableImplV2.java

```diff
@@ -0,0 +1,590 @@
+      byteBuffer.get(schemaBytes);
+      _dataSchema = DataSchema.fromBytes(schemaBytes);
+      _columnOffsets = computeColumnOffsets();
+    }
+
+    // Read fixed size data.
+    if (fixedSizeDataLength != 0) {
+      _fixedSizeDataBytes = new byte[fixedSizeDataLength];
+      byteBuffer.position(fixedSizeDataStart);
+      byteBuffer.get(_fixedSizeDataBytes);
+      _fixedSizeData = ByteBuffer.wrap(_fixedSizeDataBytes);
+    }
+
+    // Read variable size data.
+    if (variableSizeDataLength != 0) {
+      _variableSizeDataBytes = new byte[variableSizeDataLength];
+      byteBuffer.position(variableSizeDataStart);
+      byteBuffer.get(_variableSizeDataBytes);
+      _variableSizeData = ByteBuffer.wrap(_variableSizeDataBytes);
+    }
+  }
+
+  private int[] computeColumnOffsets() {
+    int[] columnOffsets = new int[_numColumns];
+    for (int i = 0; i < _numColumns; i++) {
+      columnOffsets[i] = _rowSizeInBytes;
+      FieldSpec.DataType dataType = _dataSchema.getColumnType(i);
+      switch (dataType) {
+        case BOOLEAN:
+          _rowSizeInBytes += 1;
+          break;
+        case BYTE:
+          _rowSizeInBytes += 1;
+          break;
+        case CHAR:
+          _rowSizeInBytes += 2;
+          break;
+        case SHORT:
+          _rowSizeInBytes += 2;
+          break;
+        case INT:
+          _rowSizeInBytes += 4;
+          break;
+        case LONG:
+          _rowSizeInBytes += 8;
+          break;
+        // TODO: fix float size (should be 4).
+        case FLOAT:
+          _rowSizeInBytes += 8;
+          break;
+        case DOUBLE:
+          _rowSizeInBytes += 8;
+          break;
+        case STRING:
+          _rowSizeInBytes += 4;
+          break;
+        // Object and array. ()
+        default:
+          _rowSizeInBytes += 8;
+          break;
+      }
+    }
+    return columnOffsets;
+  }
+
+  @SuppressWarnings("ResultOfMethodCallIgnored")
+  private Map<String, Map<Integer, String>> deserializeDictionaryMap(byte[] bytes)
+      throws IOException {
+    Map<String, Map<Integer, String>> dictionaryMap = new HashMap<>();
+    ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(bytes);
+    DataInputStream dataInputStream = new DataInputStream(byteArrayInputStream);
+    int numDictionaries = dataInputStream.readInt();
+    for (int i = 0; i < numDictionaries; i++) {
+      int columnNameLength = dataInputStream.readInt();
+      byte[] columnNameBytes = new byte[columnNameLength];
+      dataInputStream.read(columnNameBytes);
+      Map<Integer, String> dictionary = new HashMap<>();
+      dictionaryMap.put(new String(columnNameBytes, UTF_8), dictionary);
+      int dictionarySize = dataInputStream.readInt();
+      for (int j = 0; j < dictionarySize; j++) {
+        int key = dataInputStream.readInt();
+        int valueLength = dataInputStream.readInt();
+        byte[] valueBytes = new byte[valueLength];
+        dataInputStream.read(valueBytes);
```

Comment:
If you use the result of line 211 to assert that number bytes read == valueLength, then you won't need the SuppressWarning in line 193. But I'll leave it up to you.